File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.size();
        unordered_map<char,int> mp; // faster than map
        int j = 0, max_len = 0;

        for (int i = 0; i < n; i++) {
            mp[s[i]]++;

            while (mp[s[i]] > 1) { // duplicate found
                mp[s[j]]--;
                if (mp[s[j]] == 0) {
                    mp.erase(s[j]);
                }
                j++;
            }
            max_len = max(max_len, i - j + 1);
        }

        return max_len;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        from collections import defaultdict

        freq = defaultdict(int)
        left = 0
        max_len = 0

        for right in range(len(s)):
            freq[s[right]] += 1

            while freq[s[right]] > 1:
                freq[s[left]] -= 1
                if freq[s[left]] == 0:
                    del freq[s[left]]
                left += 1
                 
            
            max_len = max(max_len,right-left+1)
        
        return max_len

                

        